sdb/commands/linux/internal/slub_helpers.py

```diff
@@ -122,3 +124,174 @@ def util(cache: drgn.Object) -> int:
+    padding = 0
+    if cache.flags & SLAB_RED_ZONE_FLAG:
+        padding += int(cache.red_left_pad)
+    return padding
+
+
+def cache_get_free_pointer(cache: drgn.Object, p: drgn.Object) -> drgn.Object:
+    """
+    Get the next pointer in the freelist. Note, that this
+    function assumes that CONFIG_SLAB_FREELIST_HARDENED
+    is set in the target
+    """
+    assert cache.type_.type_name() == 'struct kmem_cache *'
+    assert p.type_.type_name() == 'void *'
+    hardened_ptr = p + cache.offset.value_()
+
+    #
+    # We basically do what `freelist_dereference()` and
+    # `freelist_ptr()` do in the kernel source:
+    #
+    # ptr <- (void *)*(unsigned long *)(hardened_ptr)
+    #
+    intermediate_ulong = drgn.Object(sdb.get_prog(),
+                                     type='unsigned long',
+                                     address=hardened_ptr.value_())
+    ptr = drgn.Object(sdb.get_prog(),
+                      type='void *',
+                      value=intermediate_ulong.value_())
+
+    #
+    # ptr_addr <- (unsigned long)hardened_ptr
+    #
+    ptr_addr = drgn.Object(sdb.get_prog(),
+                           type='unsigned long',
+                           value=hardened_ptr.value_())
+
+    #
+    # return (void *)((unsigned long)ptr ^ cache->random ^ ptr_addr)
+    #
+    ptr_as_ulong = drgn.Object(sdb.get_prog(),
+                               type='unsigned long',
+                               value=ptr.value_())
+    clean_ptr_val = ptr_as_ulong.value_()
+    clean_ptr_val ^= cache.random.value_()
+    clean_ptr_val ^= ptr_addr.value_()
+    return drgn.Object(sdb.get_prog(), type='void *', value=clean_ptr_val)
+
+
+def cache_get_free_set(cache: drgn.Object, page: drgn.Object) -> Set[int]:
+    """
+    Return a set of addresses that point to all the freed
+    objects in this slab.
+    """
+
+    def __cache_get_free_set_gen() -> Iterable[int]:
+        p = page.freelist
+        while p.value_() != 0x0:
+            yield p.value_()
+            p = cache_get_free_pointer(cache, p)
+
+    return set(__cache_get_free_set_gen())
+
+
+def cache_gather_children(cache: drgn.Object) -> Set[int]:
+    return {
+        child.value_() for child in list_for_each_entry(
+            "struct kmem_cache", cache.memcg_params.children.address_of_(),
+            "memcg_params.children_node")
+    }
+
+
+def cache_get_partials(cache: drgn.Object) -> Set[int]:
+    assert cache.type_.type_name() == 'struct kmem_cache *'
+    partials = []
+
+    node = cache.node[0].partial  # assumption nr_node_ids == 0
+    for page in list_for_each_entry("struct page", node.address_of_(), "lru"):
+        partials.append(page.value_())
+
+    if not is_root_cache(cache):
+        return set(partials)
+
+    for child in list_for_each_entry("struct kmem_cache",
+                                     cache.memcg_params.children.address_of_(),
+                                     "memcg_params.children_node"):
```

Comment:
It would be nice if this could use reuse the list_for_each_entry call that's already in cache_gather_children().  Maybe if we changed that to be a generator that yields the drgn.Object's, and then the other caller could create the set from that?